# 线性代数问题的数值解

科学计算理论的背后几乎离不开线性代数的计算问题，如矩阵乘法、矩阵的逆、矩阵分解、特征值和特征向量等． NumPy库可以解决线性代数相关的计算，只不过需要调用 NumPy 的子模块 linalg （线性代数的缩写），该模块几乎提供了线性代数所需的所有功能．

强调一下，在 NumPy 库的 array 数组中，＊和 multiply 运算等价，都表示矩阵
的逐个元素相乘＠和dot() 函数表示矩阵乘法

### 1. 向量和矩阵的运算

 形状为 (n,）的一维数组既可以看成行向量，又可以看成列向量，因而本例中 inner(a,b) 与矩阵乘法 dot(a,b) 等价

$$
\text{已知 } \boldsymbol{\alpha} = [1, 2, 3]^T, \boldsymbol{\beta} = [4, 5, 6]^T, \text{ 求 } \|\boldsymbol{\alpha}\|_2 \text{（向量 } \boldsymbol{\alpha} \text{ 的长度）}, \boldsymbol{\alpha} \cdot \boldsymbol{\beta}, \boldsymbol{\alpha} \times \boldsymbol{\beta}. 
$$

In [4]:
from numpy import arange, cross, inner
from numpy.linalg import norm
a = arange(1, 4)  # 行向量（一维数组）
b = arange(4, 7)  # 行向量
print(f"a的二范式(模)为：{norm(a)}")
print(f"a点乘b={a.dot(b)}")
print(f"axb的内积={inner(a, b)}")  # 与上面等价
print(f"a叉乘b={cross(a, b)}")

a的二范式(模)为：3.7416573867739413
a点乘b=32
axb的内积=32
a叉乘b=[-3  6 -3]


$$
\text{已知 } \boldsymbol{A} = \begin{bmatrix}
1 & 2 & 3 & 4 \\
5 & 6 & 7 & 8 \\
9 & 10 & 11 & 12 \\
13 & 14 & 15 & 16
\end{bmatrix}, \boldsymbol{B} = \begin{bmatrix}
1 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1
\end{bmatrix}, \text{ 求 } |\boldsymbol{A}|, \text{ } \boldsymbol{A} \text{ 的秩 } R(\boldsymbol{A}), \text{ } \boldsymbol{A}^T, \text{ } (\boldsymbol{A} + 10\boldsymbol{E})^{-1}, \text{ } \boldsymbol{A}^2, \text{ } \boldsymbol{A}\boldsymbol{B}, \text{ 构造分块矩阵 } [\boldsymbol{A}, \boldsymbol{B}], \begin{bmatrix} \boldsymbol{A} \\ \boldsymbol{B} \end{bmatrix}, \boldsymbol{A}_1 = \begin{bmatrix} 1 & 2 \\ 5 & 6 \end{bmatrix} \text{（提出 } \boldsymbol{A} \text{ 的左上角子块构成的矩阵）}, \boldsymbol{A}_2 = \begin{bmatrix} 1 & 2 & 3 & 4 \\ 5 & 6 & 7 & 8 \\ 9 & 10 & 11 & 12 \end{bmatrix} \text{（删除 } \boldsymbol{A} \text{ 的第 4 行得到的矩阵）}.
$$

In [14]:
import numpy as np
import numpy.linalg as LA
A = np.arange(1, 17).reshape(4, 4)
B = np.eye(4)
print(f"A的行列式为：{LA.det(A)}")
print(f"A的秩为：{LA.matrix_rank(A)}")
print(f"A的转置矩阵为：\n{A.transpose()}")
print(f"所求矩阵的逆矩阵为：{LA.inv(A+10*B)}")
print(f"A的平方为：{A**2}")
print(f"A,B乘积为：{A.dot(B)}")
print(f"横联分块矩阵为：{np.c_[A, B]}")
print(f"纵联分块矩阵为为：{np.r_[A, B]}")
print(f"左上角分区矩阵：\n{A[0:2, 0:2]}")
print(f"删除A的第4行得（方法一）：{A[0:3, :]}")  # 没有改变原来的矩阵，只是输出了其前三行 0:3 0，1，2（3到不了）
A1 = A.copy()
A1 = np.delete(A1, 3, axis=0)  # axis=0行操作，axis=1列操作
print(f"删除A的第4行得（方法二）：\n{A1}")

A的行列式为：-1.820448242817726e-31
A的秩为：2
A的转置矩阵为：
[[ 1  5  9 13]
 [ 2  6 10 14]
 [ 3  7 11 15]
 [ 4  8 12 16]]
所求矩阵的逆矩阵为：[[ 0.11277778  0.00333333 -0.00611111 -0.01555556]
 [-0.005       0.09       -0.015      -0.02      ]
 [-0.02277778 -0.02333333  0.07611111 -0.02444444]
 [-0.04055556 -0.03666667 -0.03277778  0.07111111]]
A的平方为：[[  1   4   9  16]
 [ 25  36  49  64]
 [ 81 100 121 144]
 [169 196 225 256]]
A,B乘积为：[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]
 [13. 14. 15. 16.]]
横联分块矩阵为：[[ 1.  2.  3.  4.  1.  0.  0.  0.]
 [ 5.  6.  7.  8.  0.  1.  0.  0.]
 [ 9. 10. 11. 12.  0.  0.  1.  0.]
 [13. 14. 15. 16.  0.  0.  0.  1.]]
纵联分块矩阵为为：[[ 1.  2.  3.  4.]
 [ 5.  6.  7.  8.]
 [ 9. 10. 11. 12.]
 [13. 14. 15. 16.]
 [ 1.  0.  0.  0.]
 [ 0.  1.  0.  0.]
 [ 0.  0.  1.  0.]
 [ 0.  0.  0.  1.]]
左上角分区矩阵：
[[1 2]
 [5 6]]
删除A的第4行得（方法一）：[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
删除A的第4行得（方法二）：
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


### 齐次线性方程组的的数值解

使用scipy.linalg的 null_space 函数，可以求齐次方程组的基础解系

| 参数名               | 类型           | 默认值     | 说明                                                                                  |
| ----------------- | ------------ | ------- | ----------------------------------------------------------------------------------- |
| **A**             | `array_like` | –       | 待求零空间的 **数值矩阵**（形状 `(m, n)`）。                                                       |
| **rcond**         | `float`      | `None`  | 奇异值阈值：小于 `rcond * σ_max` 的奇异值视为零。<br>若设为 `None`，则使用 `max(m, n) * eps`（`eps` 为机器精度）。 |
| **overwrite\_a**  | `bool`       | `False` | 是否允许就地覆盖输入矩阵 `A` 以节省内存。                                                             |
| **check\_finite** | `bool`       | `True`  | 是否检查输入矩阵只含有限数；设 `False` 可跳过检查以提速。                                                   |

| 返回值   | 类型                          | 说明                                                                                 |
| ----- | --------------------------- | ---------------------------------------------------------------------------------- |
| **Z** | `numpy.ndarray` 形状 `(n, k)` | 零空间的一组 **标准正交基**（每列是一个向量，满足 `A @ Z ≈ 0`）。<br>若矩阵满秩，则 `k = 0`，返回空数组 `shape=(n, 0)`。 |




| 对比维度  | sympy.Matrix.nullspace()    | scipy.linalg.null\_space()           |
| ----- | --------------------------- | ------------------------------------ |
| 所属库   | SymPy（符号数学）                 | SciPy（数值计算）                          |
| 数据类型  | `sympy.Matrix`（符号/任意精度）     | `numpy.ndarray`（双精度浮点）               |
| 数值/符号 | 精确符号计算，可处理符号表达式             | 纯数值计算，输入必须可转成浮点                      |
| 算法    | 基于高斯-约旦消元（RREF）             | 基于 SVD（奇异值分解）                        |
| 容差参数  | 无专门 `tol`，用自定义 `iszerofunc` | 有关键字 `tol`（默认 ≈1e-13）控制零奇异值截断        |
| 返回值类型 | `list[sympy.Matrix]`（列向量列表） | `numpy.ndarray` 形状 `(n, k)`，每列是一组正交基 |
| 正交归一化 | 默认不单位化；可选 `normalize=True`  | 返回的基已正交且单位化（SVD 自然保证）                |
| 性能    | 适合小规模、符号/精确需求               | 适合大规模稠密/稀疏数值矩阵                       |
| 典型用途  | 符号推导、精确零空间                  | 数值线性代数、机器学习预处理                       |

求下列齐次线性方程组的基础解系

$$
\begin{cases}
x_1 - 5x_2 + 2x_3 - 3x_4 = 0\\
5x_1 + 3x_2 + 6x_3 - x_4 = 0\\
2x_1 + 4x_2 + 2x_3 + x_4 = 0
\end{cases}
$$

In [15]:
# 输出为浮点数值
import numpy as np
from scipy.linalg import null_space
A = np.array([[1, -5, 2, -3], [5, 3, 6, -1], [2, 4, 2, 1]])
print(f"A的零空间（即基础解系）为：{null_space(A)}")

A的零空间（即基础解系）为：[[-0.27094291  0.75074375]
 [ 0.41266727 -0.01611393]
 [-0.12400881 -0.64280109]
 [-0.86076563 -0.15142959]]


### 3. 非齐次线性方程组的数值解

$$
\textbf{定理 3.1} \quad n \text{ 元线性方程组 } \boldsymbol{A}\boldsymbol{x} = \boldsymbol{b}, \\
\begin{aligned}
&(1) \text{ 无解的充分必要条件是 } R(\boldsymbol{A}) < R(\boldsymbol{A}, \boldsymbol{b}); \\
&(2) \text{ 有唯一解的充分必要条件是 } R(\boldsymbol{A}) = R(\boldsymbol{A}, \boldsymbol{b}) = n; \\
&(3) \text{ 有无穷多解的充分必要条件是 } R(\boldsymbol{A}) = R(\boldsymbol{A}, \boldsymbol{b}) < n.
\end{aligned} \\
\text{无论数学上 } \boldsymbol{A}\boldsymbol{x} = \boldsymbol{b} \text{ 是否存在解，或者是否存在多解，Python 的求解命令 } \\
\boldsymbol{x} = \text{pinv}(\boldsymbol{A}) \cdot \text{dot}(\boldsymbol{b}) \text{ 总是给出唯一解，给出解的情况如下：} \\
\begin{aligned}
&(1) \text{ 当方程组有无穷多解时，Python 给出的是最小范数解.} \\
&(2) \text{ 当方程组无解时，Python 给出的是最小二乘解 } \boldsymbol{x}^*, \text{ 所谓的最小二乘解 } \boldsymbol{x}^* \\
&\text{ 是满足 } \|\boldsymbol{A}\boldsymbol{x}^* - \boldsymbol{b}\|^2 \text{ 最小的解，即方程两边误差平方和最小的解.}
\end{aligned}
$$

#### （1）唯一解情形

求下列非线性方程组的数值解

$$
\begin{cases}
2x_1 + x_2 - 5x_3 + x_4 = 8, \\
x_1 - 3x_2 - 6x_4 = 6, \\
2x_2 - x_3 + 2x_4 = -2, \\
x_1 + 4x_2 - 7x_3 + 6x_4 = 2.
\end{cases}
$$

记上述线性方程组为 $Ax=b$，可以验证系数矩阵 $A$ 的秩$R(A) = 4$，所
以线性方程组有唯一解$x＝A^-1b$（两边同时左乘$A^-1$可求解x），利用 Python 软件求得
$$ x=[4, -1.5556， -0.2222， 0.4444]^T $$

In [22]:
import numpy as np
import scipy.linalg as LA
A = np.array([[2, 1, -5, 1], [1, -3, 0, -6], [0, 2, -1, 2], [1, 4, -7, 6]])
b = np.array([8, 6, -2, 2])
b.transpose()
# np.transpose(b)
print(f"系数矩阵的秩为：{np.linalg.matrix_rank(A)}")
print(f"线性方程组的唯一解为：{LA.inv(A).dot(b)}")  # 使用逆矩阵
print(f"线性方程组的唯一解为：{LA.pinv(A).dot(b)}")  # 使用伪逆
print(f"线性方程组的唯一解为：{LA.solve(A, b)}")  # 利用solve求解

系数矩阵的秩为：4
线性方程组的唯一解为：[ 4.         -1.55555556 -0.22222222  0.44444444]
线性方程组的唯一解为：[ 4.         -1.55555556 -0.22222222  0.44444444]
线性方程组的唯一解为：[ 4.         -1.55555556 -0.22222222  0.44444444]


#### (2)多解情形

下列非齐次线性方程组有无穷多解，求它的最小范数解

$$
\begin{cases}
x_1 + x_2 - 3x_3 - x_4 = 1, \\
3x_1 - x_2 - 3x_3 + 4x_4 = 4, \\
x_1 + 5x_2 - 9x_3 - 8x_4 = 0.
\end{cases}
$$

In [24]:
import numpy as np
import scipy.linalg as LA
A = np.array([[1, 1, -3, -1], [3, -1, -3, 4], [1, 5, -9, -8]])
b = np.array([1, 4, 0])
b.reshape(3, 1)
print(f"该方程的最小范数解为：{LA.pinv(A).dot(b)}")

该方程的最小范数解为：[ 0.35040431 -0.0916442  -0.38814016  0.42318059]


#### (3)无解情形

求下列矛盾方程的最小二乘解（无解）

$$
\begin{cases}
x_1 + x_2 = 1\\
2x_1 + 2x_2 = 3\\
x_1 + 2x_2 = 2
\end{cases}
$$

In [25]:
from numpy import array
from numpy.linalg import pinv
A = array([[1, 1], [2, 2], [1, 2]])
b = array([1, 3, 2])
x = pinv(A).dot(b)
print(f"该无解方程的最小二乘解为：{x}")

该无解方程的最小二乘解为：[0.8 0.6]


#### 4. 特征值和特征向量

求下列矩阵的特征值和特征向量

$$
A = \begin{bmatrix}
0 & -2 & 2 \\
-2 & -3 & 4 \\
2 & 4 & -3
\end{bmatrix}
$$

In [27]:
# 数值解
import numpy as np
from numpy.linalg import eig
A = np.array([[0, -2, 2], [-2, -3, 4], [2, 4, -3]])
values, vectors = eig(A)
print(f"A的特征值为{values}\n特征向量为{vectors}")

A的特征值为[ 1. -8.  1.]
特征向量为[[ 0.94280904 -0.33333333  0.06541736]
 [-0.23570226 -0.66666667  0.68904825]
 [ 0.23570226  0.66666667  0.72175693]]


对于矩阵

$$
\boldsymbol{A} = \begin{bmatrix}
0 & -2 & 2 \\
-2 & -3 & 4 \\
2 & 4 & -3
\end{bmatrix}, \quad
\boldsymbol{P} = \begin{bmatrix}
0.9428 & -0.3333 & -0.4254 \\
-0.2357 & -0.6667 & 0.7454 \\
0.2357 & 0.6667 & 0.5247
\end{bmatrix}, \quad
\boldsymbol{\Lambda} = \begin{bmatrix}
1 & 0 & 0 \\
0 & -8 & 0 \\
0 & 0 & 1
\end{bmatrix}, \\
\text{验证 } \boldsymbol{P}^{-1}\boldsymbol{A}\boldsymbol{P} = \boldsymbol{\Lambda}.
$$

In [28]:
from numpy import array, dot
from numpy.linalg import eig, inv
A = array([[0, -2, 2], [-2, -3, 4], [2, 4, -3]])
values, vectors = eig(A)
check = dot(inv(vectors), A).dot(vectors)
print(f"check = \n{check}")

check = 
[[ 1.00000000e+00 -6.52328509e-16 -3.95551496e-17]
 [-2.13578642e-16 -8.00000000e+00 -1.61160385e-15]
 [ 6.52332138e-17  7.92611538e-16  1.00000000e+00]]
